---
title: "CMSE Project"
output: html_document
---

```{r}
## read in the SNP position file and the gene location file
snp_location_file = read.csv("snp_position.csv", header = T)
gene_location_file = read.csv("expression_location.csv", header = T)

```



```{r}
## Match SNP and gene according to their position

for(chromosome in 1:16){
  gene_chr = gene_location_file[which(gene_location_file[,2] == chromosome),]
  snp_chr = snp_location_file[which(snp_location_file[,2] == chromosome),]
  for(snp in 1:dim(snp_chr)[1]){
    for(gene in 1:dim(gene_chr)[1]){
      if(gene_chr[gene,3] <= snp_chr[snp, 3] && 
         gene_chr[gene,4] >= snp_chr[snp, 3]){
        row = matrix(c(gene_chr[gene, 1:2],
                       as.character(snp_chr[snp, 1]),snp_chr[snp, 3]),
                     nrow=1)
        write.table(row, file ="ID_Chr_SNP.txt",
                    col.names = F,row.names = F, append=T, eol="\t\n")
      
      }
      else if(gene_chr[gene,3] >= snp_chr[snp, 3] && 
           gene_chr[gene,4] <= snp_chr[snp, 3]){
        row = matrix(c(gene_chr[gene, 1:2],
                       as.character(snp_chr[snp, 1]),snp_chr[snp, 3]),
                     nrow=1)
        
        write.table(row, file ="ID_Chr_SNP.txt",
                    col.names = F,row.names = F, append=T, eol="\t\n")
      }
      
    }

  }
    
}



```




```{r}
## check how many genes have SNP
## the snp ID is not unique, add the line number of the snp in the SNP position file as its ID
ID_Chr_SNP = read.csv("ID_Chr_SNP.csv", header=F)
for(SNP_ID in 1:dim(ID_Chr_SNP)[1]){
  line = which(as.character(snp_location_file[,1]) == ID_Chr_SNP[SNP_ID,3] &
                 as.character(snp_location_file[,3]) ==
                 ID_Chr_SNP[SNP_ID,4]) + 1
  LINE = c(line, ID_Chr_SNP[SNP_ID,])
  write.table(LINE, file ="LINE_ID_SNP.txt",
              col.names = F,row.names = F,quote = F, append=T, eol="\t\n")
}

        
```



```{r}
## Read in the matrix_genotype data, expression matrix data

LINE_ID_SNP = read.csv("LINE_ID_SNP.csv", header = F)

matrix_genotype = read.csv("matrix_genotypes.csv", header = T)
expression_data= read.csv("expression_matrix.csv", header = T)
```



```{r}
## delete rows that have more than half number of NULL, null expression value provides no information

expression_matrix = expression_data[apply(expression_data =='NULL', 1,
                                          sum) < 56,]
dim(expression_matrix)
```


```{r}
## Get the genes that have no siginificantly differential expression given certain SNPs condotions (0, 1, 2) as the candidate TF

for(get_row in 1:dim(LINE_ID_SNP)[1]){
  gene.row = which(expression_matrix[,1] == LINE_ID_SNP[get_row,2])
  genotype = matrix_genotype[LINE_ID_SNP[get_row,1]-1,]
  sample.snp0 = colnames(genotype[,which(genotype == 0)])
  sample.snp1 = colnames(genotype[,which(genotype == 1)])
  sample.snp2 = colnames(genotype[,which(genotype == 2)])
  
  ## get the gene expression value according to different SNP condition
  expr.snp0 = as.numeric(as.matrix(expression_matrix[gene.row,
                                                     sample.snp0]))
  expr.snp1 = as.numeric(as.matrix(expression_matrix[gene.row,
                                                     sample.snp1]))
  expr.snp2 = as.numeric(as.matrix(expression_matrix[gene.row,
                                                     sample.snp2]))
  
  
  ## do t-test and output the indifferentilally expressed gene ID, snp line      number the position file snp ID as the candidate TFs to do subsequent       analysis
  if(length(expr.snp0) > 10 && length(expr.snp1) >10 &&
     length(expr.snp2) >10){
    t_test0_1 = t.test(x=expr.snp0, y=expr.snp1)
    t_test0_2 = t.test(x=expr.snp0, y=expr.snp2)
    t_test1_2 = t.test(x=expr.snp1, y=expr.snp2)
    if(t_test0_1[3] > 0.01 && t_test0_2[3] > 0.01 && t_test1_2[3] > 0.01){
      write.table(matrix(c(LINE_ID_SNP[get_row,2],
                           LINE_ID_SNP[get_row,1],
                           as.character(LINE_ID_SNP[get_row,4])),nrow = 1),
                  file = "indifferent_expressed_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
    }
  } else if(length(expr.snp0) >10  && length(expr.snp1) >10 &&
            length(expr.snp2) <10){
    t_test0_1 = t.test(x=expr.snp0, y=expr.snp1)
    if(t_test0_1[3] > 0.01){
      write.table(matrix(c(LINE_ID_SNP[get_row,2],
                           LINE_ID_SNP[get_row,1],
                           as.character(LINE_ID_SNP[get_row,4])),nrow = 1),
                  file = "indifferent_expressed_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
    }
  } else if(length(expr.snp0) >10  && length(expr.snp2) >10 &&
            length(expr.snp1) < 10){
    t_test0_2 = t.test(x=expr.snp0, y=expr.snp2)
    if(t_test0_2[3] > 0.01){
      write.table(matrix(c(LINE_ID_SNP[get_row,2],
                           LINE_ID_SNP[get_row,1],
                           as.character(LINE_ID_SNP[get_row,4])),nrow = 1),
                  file = "indifferent_expressed_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
        
    }
    
  } else if(length(expr.snp1) >10  && length(expr.snp2) >10 &&
            length(expr.snp0) < 10){
    t_test1_2 = t.test(x=expr.snp1, y=expr.snp2)
    if(t_test1_2[3] > 0.01){
      write.table(matrix(c(LINE_ID_SNP[get_row,2],
                           LINE_ID_SNP[get_row,1],
                           as.character(LINE_ID_SNP[get_row,4])),nrow = 1),
                  file = "indifferent_expressed_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")              
    }
  }

} 

```



```{r}
## Get the intersection of the genes that code protein and the genes in the    expression file
## read in the id_to_gene_name_no_unknown

id_to_gene = read.csv("id_to_gene_name_no_unknowns.csv", header = T)
ID.id_to_gene = id_to_gene[,1]
ID.expression_matrix = expression_matrix[,1]

## extract the common gene from the expression matrix
common_gene = expression_matrix[intersect(ID.id_to_gene,
                                          ID.expression_matrix),]

```



```{r}
## read in the indif_expr_gene file
indif_expr_gene = read.csv("indifferent_expressed_gene.csv",
                                      header = F)
```



```{r}
## do t-test for every gene given one SNP in the indifferent_expressed_gene
for(indif_snp in 1:dim(indif_expr_gene)[1]){

  genotype2 = matrix_genotype[indif_expr_gene[indif_snp,2]-1,]
  ## get the sample ID given SNP conditions (0, 1, 2)
  indif_snp0 = colnames(genotype2[,which(genotype2 == 0)])
  indif_snp1 = colnames(genotype2[,which(genotype2 == 1)])
  indif_snp2 = colnames(genotype2[,which(genotype2 == 2)])
  
  # get gene (target) expression value according to different conditions
  for(com_gen.row in 1:dim(common_gene)[1]){
   expr.indif_snp0 = as.numeric(as.matrix(common_gene[com_gen.row,
                                                      indif_snp0]))
   expr.indif_snp1 = as.numeric(as.matrix(common_gene[com_gen.row,
                                                      indif_snp1]))
   expr.indif_snp2 = as.numeric(as.matrix(common_gene[com_gen.row,
                                                      indif_snp2]))
   ## do the t-test, if p-value < 0.01, output the candidate TF ID, SNP ID        targets ID, p-value 
   if(length(expr.indif_snp0) > 20 && length(expr.indif_snp1) >20 &&
      length(expr.indif_snp2) >20){
    
     indif_t_test0_1 = t.test(x=expr.indif_snp0, y=expr.indif_snp1)
     indif_t_test0_2 = t.test(x=expr.indif_snp0, y=expr.indif_snp2)
     indif_t_test1_2 = t.test(x=expr.indif_snp1, y=expr.indif_snp2)
    
     if(indif_t_test0_1[3] < 0.01 && indif_t_test0_2[3] < 0.01 &&
        indif_t_test1_2[3] < 0.01){
       write.table(matrix(c(indif_expr_gene[indif_snp,1],
                            indif_expr_gene[indif_snp,2],
                           as.character(indif_expr_gene[indif_snp,3]),
                           common_gene[com_gen.row,1],
                           as.numeric(indif_t_test0_2[3])),
                         nrow=1),
                  file = "target_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
       next
     }
   }else if(length(expr.indif_snp0) > 20 && length(expr.indif_snp1) >20 &&
             length(expr.indif_snp2) <20){
     if(sum(is.na(expr.indif_snp0)) == length(expr.indif_snp0) ||
        sum(is.na(expr.indif_snp1)) == length(expr.indif_snp1))
     {
       next
     } else{
       indif_t_test0_1 = t.test(x=expr.indif_snp0, y=expr.indif_snp1)
       if(indif_t_test0_1[3] < 0.01){
       write.table(matrix(c(indif_expr_gene[indif_snp,1],
                            indif_expr_gene[indif_snp,2],
                           as.character(indif_expr_gene[indif_snp,3]),
                           common_gene[com_gen.row,1],
                           as.numeric(indif_t_test0_1[3])),
                         nrow=1),
                  file = "target_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
       next
     }
     } 
   } else if(length(expr.indif_snp0) > 20 && length(expr.indif_snp2) >20 &&
             length(expr.indif_snp1) <20){
     if(sum(is.na(expr.indif_snp0)) == length(expr.indif_snp0) ||
        sum(is.na(expr.indif_snp2)) == length(expr.indif_snp2))
     {
       next
     } else
       indif_t_test0_2 = t.test(x=expr.indif_snp0, y=expr.indif_snp2)
       if(indif_t_test0_2[3] < 0.01){
       write.table(matrix(c(indif_expr_gene[indif_snp,1],
                            indif_expr_gene[indif_snp,2],
                           as.character(indif_expr_gene[indif_snp,3]),
                           common_gene[com_gen.row,1],
                           as.numeric(indif_t_test0_2[3])),
                         nrow=1),
                  file = "target_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
       next
     }
   } else if(length(expr.indif_snp1) > 20 && length(expr.indif_snp2) >20 &&
             length(expr.indif_snp0) <20){
     if(sum(is.na(expr.indif_snp1)) == length(expr.indif_snp1) ||
        sum(is.na(expr.indif_snp2)) == length(expr.indif_snp2))
     {
       next
     } else
       indif_t_test1_2 = t.test(x=expr.indif_snp1, y=expr.indif_snp2)
       if(indif_t_test1_2[3] < 0.01){
       write.table(matrix(c(indif_expr_gene[indif_snp,1],
                            indif_expr_gene[indif_snp,2],
                           as.character(indif_expr_gene[indif_snp,3]),
                           common_gene[com_gen.row,1],
                           as.numeric(indif_t_test1_2[3])),
                         nrow=1),
                  file = "target_gene.txt",
                  col.names = F,row.names = F,quote=F, append=T, eol="\t\n")
       next
     
     } 
   }
   
 }
}

```



  
```{r}
## read in the target_gene file, do the analysis
target = read.csv("target_gene.csv", header = F)
```


```{r}
## adjust the p-value first
p_value = target[,5]
fdr = p.adjust(p_value, method = "fdr", n=length(p_value))
target_fdr = cbind(target, fdr)

## extract the TF candidate and the target that have fd <0.001
fdr0001 = target_fdr[which(target_fdr[,6]<0.001),]

```


```{r}
## check the non_candidate gene (candidate TF)
candidate_gene = unique(fdr0001[,1])

## check the genes that are not differentially expressed but not TF   candidate 
non_candi_gene = setdiff(unique(indif_expr_gene[,1]),candidate_gene )

```




```{r}

## read in the file RegulationMatrix and the idtoname file
## rename the rownames
RegulationMatrix = read.csv("RegulationMatrix_Documented_2013927.csv",
                            header = T)
row.names(RegulationMatrix) = RegulationMatrix[,1]
RegulationMatrix_1 = RegulationMatrix[,-1]

id_to_name = read.csv("id_to_gene_name_no_unknowns.csv")
row.names(id_to_name) = id_to_name[,1]


## get the top 5 genes with the largest number of SNPs
candi5513 = fdr0001[which(fdr0001[,1] == '5513'),]
candi78 = fdr0001[which(fdr0001[,1] == '78'),]
candi4414 = fdr0001[which(fdr0001[,1] == '4414'),]
candi336 = fdr0001[which(fdr0001[,1] == '336'),]
candi4442 = fdr0001[which(fdr0001[,1] == '4442'),]
```


```{r}
## analysis genes that have large number of SNPs in a reverse way
## check the the number of targets that a gene has
## sort the targets according to the number of TF that regulates that gene
gene_name.5513 = as.character(id_to_name['5513',][,4])
known_5513_targets = RegulationMatrix_1[,which(RegulationMatrix_1['Fzf1p',]
                                               == 1)]
sort(apply(known_5513_targets, 2, sum),decreasing = F)


gene_name.78 = as.character(id_to_name['78',][,4]) 
gene_name.78
known_78_targets = RegulationMatrix_1[,which(RegulationMatrix_1['Nsi1p',]
                                               == 1)]
sort(apply(known_78_targets, 2, sum),decreasing = F)


gene_name.4414 = as.character(id_to_name['4414',][,4])
gene_name.4414
known_4414_targets = RegulationMatrix_1[,which(RegulationMatrix_1['Rgt1p',]
                                               == 1)]
sort(apply(known_4414_targets, 2, sum),decreasing = F)

gene_name.336 = as.character(id_to_name['336',][,4])
gene_name.336
known_336_targets = RegulationMatrix_1[,which(RegulationMatrix_1['Zds1',]
                                               == 1)]
sort(apply(known_336_targets, 2, sum),decreasing = F)


gene_name.4442 = as.character(id_to_name['4442',][,4])
gene_name.4442
known_4442_targets = RegulationMatrix_1[,which(RegulationMatrix_1['Imp2',]
                                               == 1)]
sort(apply(known_4442_targets, 2, sum),decreasing = F)


```









  
  